Non-reflecting BC for linear stability analysis solvers

lee-hyeoksu opened the pull request:
## Description

Please include a summary of the changes and the related issue(s) if they exist.
Please also include relevant motivation and context.

Fixes #(issue) [optional]

### Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Something else

### Scope

- [ ] This PR comprises a set of related changes with a common goal

If you cannot check the above box, please split your PR into multiple PRs that each have a common goal.

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. 
Provide instructions so we can reproduce. 
Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

**Test Configuration**:

* What computers and compilers did you use to test this:

## Checklist

- [ ] I have added comments for the new code
- [ ] I added Doxygen docstrings to the new code
- [ ] I have made corresponding changes to the documentation (`docs/`)
- [ ] I have added regression tests to the test suite so that people can verify in the future that the feature is behaving as expected
- [ ] I have added example cases in `examples/` that demonstrate my new feature performing as expected.
They run to completion and demonstrate "interesting physics"
- [ ] I ran `./mfc.sh format` before committing my code
- [ ] New and existing tests pass locally with my changes, including with GPU capability enabled (both NVIDIA hardware with NVHPC compilers and AMD hardware with CRAY compilers) and disabled
- [ ] This PR does not introduce any repeated code (it follows the [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) principle)
- [ ] I cannot think of a way to condense this code and reduce any introduced additional line count

### If your code changes any code source files (anything in `src/simulation`)

To make sure the code is performing as expected on GPU devices, I have:
- [ ] Checked that the code compiles using NVHPC compilers
- [ ] Checked that the code compiles using CRAY compilers
- [ ] Ran the code on either V100, A100, or H100 GPUs and ensured the new feature performed as expected (the GPU results match the CPU results)
- [ ] Ran the code on MI200+ GPUs and ensure the new features performed as expected (the GPU results match the CPU results)
- [ ] Enclosed the new feature via `nvtx` ranges so that they can be identified in profiles
- [ ] Ran a Nsight Systems profile using `./mfc.sh run XXXX --gpu -t simulation --nsys`, and have attached the output file (`.nsys-rep`) and plain text results to this PR
- [ ] Ran an Omniperf profile using `./mfc.sh run XXXX --gpu -t simulation --omniperf`, and have attached the output file and plain text results to this PR.
- [ ] Ran my code using various numbers of different GPUs (1, 2, and 8, for example) in parallel and made sure that the results scale similarly to what happens if you run without the new code/feature

sbryngelson commented on src/pre_process/m_initial_condition.fpp:
```diff
@@ -394,43 +396,43 @@ contains
+            call s_instability_wave(2*pi*2.0/59.0, -2*pi*2.0/59.0, wave_tmp, shift(5))
             wave2 = wave2 + wave_tmp
-            call s_instability_wave(2*pi*1.0/Lx, -2*pi*1.0/Lz, tr, ti, wave_tmp, 2*pi*29d0/31d0)
+            call s_instability_wave(2*pi*1.0/59.0, -2*pi*1.0/59.0, wave_tmp, shift(6))
```
`59.` here (and in many other parts of this PR) serves as a "magic number." it won't be clear to the reader what it is for/what it means. I don't understand it myself. I suggest giving it a variable name that makes sense. Is it scale invariant?

sbryngelson commented on src/pre_process/m_initial_condition.fpp:
```diff
@@ -526,69 +601,201 @@ contains
-                ii = 1; jj = 5; ci((ii - 1)*(n + 1) + j, (jj - 1)*(n + 1) + k) = -rho_mean(j)*d(j, k); 
-                ii = 5; jj = 3; ci((ii - 1)*(n + 1) + j, (jj - 1)*(n + 1) + k) = -d(j, k)/rho_mean(j); 
-                ii = 3; jj = 5; ci((ii - 1)*(n + 1) + j, (jj - 1)*(n + 1) + k) = -gam*(p_mean + pi_inf)*d(j, k); 
+        do j = 0, n + 1
```
what does all of this do? please add more comments to your code.

sbryngelson commented on src/pre_process/m_initial_condition.fpp:
```diff
@@ -607,30 +814,81 @@ contains
-                        wave(5, i, j, k) = vnr(4*(n + 1) + j - 1)*cos(ang) - vni(4*(n + 1) + j - 1)*sin(ang) ! v
+                        ang = alpha*(x_cc(i)*xratio) + beta*(z_cc(k)*xratio) + shift
                     end if
+                    wave(1, i, j, k) = xcr(j)*cos(ang) - xci(j)*sin(ang) ! rho
```
where does this come from? can you include a paper reference in a comment?

sbryngelson commented on docs/documentation/case.md:
```diff
@@ -677,7 +677,7 @@ The parameters are optionally used to define initial velocity profiles and pertu
 - `vel_profile` activates setting the mean streamwise velocity to hyperbolic tangent profile. This option works only for 2D and 3D cases.
 
-- `instability_wave` activates the perturbation of initial velocity by instability waves obtained from linear stability analysis for a mixing layer with hyperbolic tangent mean streamwise velocity profile. This option only works for `n > 0`, `bc_y%[beg,end] = -5`, and ``vel_profile = 'T'``.
+- `instability_wave` activates the perturbation of initial velocity by instability waves obtained from linear stability analysis for a mixing layer with hyperbolic tangent mean streamwise velocity profile. This option only works for `n > 0`, `bc_y%[beg,end] = -6`, and ``vel_profile = 'T'``.
```
so this is only for mixing layers? if true, can you be more specific in the documentation here about what a mixing layer is, specifically? you may also want to change `instablility_wave` to something more descriptive, like `mixlayer_instability_trigger`

sbryngelson commented on docs/documentation/case.md:
```diff
@@ -677,7 +677,7 @@ The parameters are optionally used to define initial velocity profiles and pertu
 
 - `vel_profile` activates setting the mean streamwise velocity to hyperbolic tangent profile. This option works only for 2D and 3D cases.
```
this line of documentation needs more details. Hyperbolic tangent with what parameters? it also doesn't mention anything about mixing layers.

sbryngelson commented on src/pre_process/m_initial_condition.fpp:
```diff
@@ -439,82 +441,155 @@ contains
+        do while (.true.)
+
+            f0 = (Ca + 2d0/Web)*(fR0/fR)**(3d0*gam_b) - 2d0/(Web*fR) + 1d0 - Ca - fP
+            f1 = -3d0*gam_b*(Ca + 2d0/Web)*(fR0/fR)**(3d0*gam_b + 1d0) + 2d0/(Web*fR**2d0)
```
it looks like much of this code is not covered by tests, according to `codecov`

sbryngelson commented on src/pre_process/m_initial_condition.fpp:
```diff
@@ -526,69 +601,201 @@ contains
+
+    end subroutine s_solve_linear_system
+
+    subroutine s_instability_nonreflecting_subsonic_buffer_bc(ar, ai, hr, hi, rho_mean, mach)
```
yeah it seems like all of these mixing layer instability wave subroutines should be moved to their own module. there is a lot of code associated with it here and it could easily get confusing.

sbryngelson commented on src/pre_process/m_initial_condition.fpp:
```diff
@@ -526,69 +601,201 @@ contains
+
+        gr = 0d0
+        gi = 0d0
+        do ii = 1, 5
```
what is `5` representing? this is a "magic number"

sbryngelson commented on src/pre_process/m_initial_condition.fpp:
```diff
@@ -526,69 +601,201 @@ contains
+
+        hr = 0d0
+        hi = 0d0
+        do i = 0, 5*n - 5
```
5 is a magic number here

sbryngelson commented on src/pre_process/m_initial_condition.fpp:
```diff
@@ -526,69 +601,201 @@ contains
         ! Find the most unstable eigenvalue and corresponding eigenvector
         k = 0
-        do i = 1, nl - 1
+        do i = 1, 5*n - 5
```
magic number -> 5

sbryngelson commented on src/pre_process/m_initial_condition.fpp:
```diff
@@ -607,30 +814,81 @@ contains
+        end do
+
+        ! rho at BC
+        xbr(0*(n + 2) + 0) = xbr(0*(n + 2) + 1) + xbr(2*(n + 2) + 1)*rho_mean(1)*mach
```
where does the `+ 2` come from?

lee-hyeoksu replied:
`+2` or `-2` is for the insertion or deletion of boundary rows at y_beg and y_end.

sbryngelson replied:
this is a magic number as well

lee-hyeoksu replied:
I removed magic numbers. I believe there is no more magic number...

lee-hyeoksu commented:
> @lee-hyeoksu FYI the Phoenix runners will be down for a few days, the computer is under maintenance.

@sbryngelson Thank you for letting me know!

lee-hyeoksu commented:
@sbryngelson I have made several changes to the code.

- Changed parameter names from `instability_wave` to `mixlayer_perturb`, from `vel_profile` to `mixlayer_vel_profile`
- Added new parameters (`mixlayer_vel_coef` and `mix layer_domain`) to remove some magic numbers and allow more degrees of freedom. 
- Updated comments and documentation accordingly.
- All the codes related to the `mixlayer_perturb` are moved to a new file `src/pre_process/m_perturbation.fpp`. As the file name implies, I also moved two subroutines `s_perturb_sphere` and `s_perturb_surrounding_flow` to `m_perturbation.fpp.
- I added more test cases to cover the changed code.
- I added some more checking statement to `src/pre_process/m_checker.fpp` to make sure that input parameters are compatible.

And I have a question regarding the 'magic number'. In the linear stability analysis solver, I have the magic number `5` in many places. However, the code only works for the magic number `5` actually. Let me elaborate on it in a little bit more detail.

The meaning of `5` is 1 continuity eq + 3 momentum eqs + 1 energy eq. Here, no matter 2D or 3D, the solver uses 3 momentum eqs. This code only allows `num_fluids = 1` and `model_eqns = 2`. If `bubbles = T`, it computes equilibrium radius additionally but it is irrelevant to the magic number `5`.

I can declare a variable and set it as `5` to avoid the repeated use of`5`. But, in order to completely get rid of the magic number, the code would be messier because I have to replace `1`, `2`, `3`, `4`, `5` by `contxb`, `momxb`, `momxb + 1`, `momxe` and `E_idx` in many places, for example:

```
        do j = 0, n + 1
            ii = 1; jj = 1; br((ii - 1)*(n + 2) + j, (jj - 1)*(n + 2) + j) = alpha*u_mean(j); 
            ii = 1; jj = 2; br((ii - 1)*(n + 2) + j, (jj - 1)*(n + 2) + j) = alpha*rho_mean(j); 
            ii = 1; jj = 3; bi((ii - 1)*(n + 2) + j, (jj - 1)*(n + 2) + j) = -drho_mean(j); 
            ii = 1; jj = 4; br((ii - 1)*(n + 2) + j, (jj - 1)*(n + 2) + j) = beta*rho_mean(j); 
            ii = 2; jj = 2; br((ii - 1)*(n + 2) + j, (jj - 1)*(n + 2) + j) = alpha*u_mean(j); 
            ii = 2; jj = 3; bi((ii - 1)*(n + 2) + j, (jj - 1)*(n + 2) + j) = -du_mean(j); 
            ii = 2; jj = 5; br((ii - 1)*(n + 2) + j, (jj - 1)*(n + 2) + j) = alpha/rho_mean(j); 
            ii = 3; jj = 3; br((ii - 1)*(n + 2) + j, (jj - 1)*(n + 2) + j) = alpha*u_mean(j); 
            ii = 4; jj = 4; br((ii - 1)*(n + 2) + j, (jj - 1)*(n + 2) + j) = alpha*u_mean(j); 
            ii = 4; jj = 5; br((ii - 1)*(n + 2) + j, (jj - 1)*(n + 2) + j) = beta/rho_mean(j); 
            ii = 5; jj = 2; br((ii - 1)*(n + 2) + j, (jj - 1)*(n + 2) + j) = gam*(p_mean + pi_inf)*alpha; 
            ii = 5; jj = 4; br((ii - 1)*(n + 2) + j, (jj - 1)*(n + 2) + j) = gam*(p_mean + pi_inf)*beta; 
            ii = 5; jj = 5; br((ii - 1)*(n + 2) + j, (jj - 1)*(n + 2) + j) = alpha*u_mean(j); 
            do k = 0, n + 1
                ii = 1; jj = 3; ci((ii - 1)*(n + 2) + j, (jj - 1)*(n + 2) + k) = -rho_mean(j)*d(j, k); 
                ii = 3; jj = 5; ci((ii - 1)*(n + 2) + j, (jj - 1)*(n + 2) + k) = -d(j, k)/rho_mean(j); 
                ii = 5; jj = 3; ci((ii - 1)*(n + 2) + j, (jj - 1)*(n + 2) + k) = -gam*(p_mean + pi_inf)*d(j, k); 
            end do
        end do
```

What would be a clever way to deal with the magic number? Please let me know if you have any suggestions!

sbryngelson commented:
@lee-hyeoksu, a magic number is a number that shows up in code without clear context. Magic numbers are confusing because while you know what 5 means, I did not, and someone who later looks at the code would also not know. 

For example, you have 5, which represents the number of equations for a certain type of situation. A good practice is to create an integer parameter in `m_constants` that represents this, set it `=5` in the code (within the parameter construct so it is known at compile time), and then use that variable instead of the number 5 throughout the code where needed.

lee-hyeoksu commented:
@sbryngelson I see. I got rid of the `5`. Please let me know if you have any better idea to make it more concise and readable.

Btw, I got two failed CI tests. One of them is `Github (ubuntu, mpi, debug, false)` failed with exit code 2. When I googled it, I found the following: "Exit code 2 signifies invalid usage of some shell built-in command. Examples of built-in commands include alias, echo, and printf. Alternatively, it could mean that we're trying to access a file or directory that doesn't exist or requires permissions." I am not sure what is causing this issue and why this happens only for `Github (ubuntu, mpi, debug, false)` test. Do you have any idea?

Also, Frontier failed to compile MFC with the error message: 

```
CMake Error at CMakeLists.txt:424 (message):
  OpenACC + Fortran is unsupported.
Call Stack (most recent call first):
  CMakeLists.txt:525 (MFC_SETUP_TARGET)
```

Interestingly, just two commits before (`93d7fbb`), Frontier test compiled successfully (and passed!). I am not sure why suddenly failed to compile.

lee-hyeoksu commented:
Oh never mind, `Github (ubuntu, mpi, debug, false)` test passed.

lee-hyeoksu commented:
I don't know why but `Github (ubuntu, mpi, debug, false)` test fails randomly with exit code 2. Could you rerun the test? @sbryngelson

sbryngelson commented on src/pre_process/m_perturbation.fpp:
```diff
@@ -0,0 +1,616 @@
+                    perturb_alpha = q_prim_vf(E_idx + perturb_flow_fluid)%sf(i, j, k)
+                    ! IF (perturb_alpha == 1d0) THEN
+                    ! Perturb partial density
+                    !    CALL RANDOM_NUMBER(rand_real)
```
should you delete these lines

lee-hyeoksu replied:
Same as the comment below. 

_`s_perturb_sphere` and `s_perturb_surrounding_flow` are not written by myself. I guess they were written many years ago. I just moved these subroutines to `m_perturbation.fpp` along with instability wave subroutines because it seems more appropriate to put the "perturbation" stuff together in a separate module. I don't think any of the example cases use these subroutines. Maybe these subroutines can be removed then?_

sbryngelson replied:
I just mean delete the commented out lines

lee-hyeoksu replied:
deleted!

sbryngelson commented on src/pre_process/m_perturbation.fpp:
```diff
@@ -0,0 +1,616 @@
+                    q_prim_vf(mom_idx%beg)%sf(i, j, k) = (1.d0 + rand_real)*q_prim_vf(mom_idx%beg)%sf(i, j, k)
+                    q_prim_vf(mom_idx%end)%sf(i, j, k) = rand_real*q_prim_vf(mom_idx%beg)%sf(i, j, k)
+                    if (bubbles) then
+                        q_prim_vf(alf_idx)%sf(i, j, k) = (1.d0 + rand_real)*q_prim_vf(alf_idx)%sf(i, j, k)
```
it looks like this bubbles code wasn't covered by the tests?

lee-hyeoksu replied:
Umm actually there are covered. It is confirmed on the codedev report.

lee-hyeoksu replied:
Same as below. 

_`s_perturb_sphere` and `s_perturb_surrounding_flow` are not written by myself. I guess they were written many years ago. I just moved these subroutines to `m_perturbation.fpp` along with instability wave subroutines because it seems more appropriate to put the "perturbation" stuff together in a separate module. I don't think any of the example cases use these subroutines. Maybe these subroutines can be removed then?_

sbryngelson commented on src/pre_process/m_perturbation.fpp:
```diff
@@ -0,0 +1,616 @@
+
+    end subroutine s_initialize_perturbation_module
+
+    subroutine s_perturb_sphere(q_prim_vf)
```
you added this subroutine but it looks like there isn't a test for it

lee-hyeoksu replied:
`s_perturb_sphere` and `s_perturb_surrounding_flow` are not written by myself. I guess they were written many years ago. I just moved these subroutines to `m_perturbation.fpp` along with instability wave subroutines because it seems more appropriate to put the "perturbation" stuff together in a separate module. I don't think any of the example cases use these subroutines. Maybe these subroutines can be removed then?

sbryngelson replied:
ah i see, nevermind!

sbryngelson commented on src/pre_process/m_perturbation.fpp:
```diff
@@ -0,0 +1,616 @@
+    use ieee_arithmetic
+
+    ! ==========================================================================
+    ! ==========================================================================
```
delete this line

sbryngelson commented:
the code looks better now but i'm a bit concerned about these debug cases failing

lee-hyeoksu commented:
@sbryngelson Yeah I guess one possible reason is time limit if there exists, because `mixlayer_perturb` takes longer than other pre_process procedures.

sbryngelson commented:
> @sbryngelson Yeah I guess one possible reason is time limit if there exists, because `mixlayer_perturb` takes longer than other pre_process procedures.

I don't think so, because it fails within 9 and 23min in the two cases I see right now.

lee-hyeoksu commented:
What does the last entity "false" mean? I will try to compile and test in debug mode and "false"

sbryngelson commented:
True/False indicates whether it is using Intel or GCC compilers (true = Intel).

lee-hyeoksu commented:
I am able to reproduce the error in debug mode. I will give it a try to fix this.

sbryngelson commented:
> I am able to reproduce the error in debug mode. I will give it a try to fix this.

Interesting - out of curiosity and for future reference, what was the full error message when you ran it?

lee-hyeoksu commented:
@sbryngelson I fixed a bug in `m_perturbation`. The error was associated with the segmentation fault. The code tried to access array index that is out of the range. The full error message is following:

```
 Processing patch           1
At line 398 of file /p/work/hyeoksu/MFC/lsa/cpu-debug/src/pre_process/m_perturbation.fpp
Fortran runtime error: Index '37' of dimension 1 of array 'rho_mean' above upper bound of 36

Error termination. Backtrace:
#0  0x14cf13449640 in ???
#1  0x14cf1344a185 in ???
#2  0x14cf1344a52a in ???
#3  0x4b7382 in __m_perturbation_MOD_s_instability_nonreflecting_subsonic_buffer_bc
	at /p/work/hyeoksu/MFC/lsa/cpu-debug/build/staging/5cfdab4a72/fypp/pre_process/m_perturbation.fpp.f90:398
#4  0x4c094b in __m_perturbation_MOD_s_solve_linear_system
	at /p/work/hyeoksu/MFC/lsa/cpu-debug/build/staging/5cfdab4a72/fypp/pre_process/m_perturbation.fpp.f90:343
#5  0x4c1d29 in __m_perturbation_MOD_s_instability_wave
	at /p/work/hyeoksu/MFC/lsa/cpu-debug/build/staging/5cfdab4a72/fypp/pre_process/m_perturbation.fpp.f90:272
#6  0x4c31e2 in __m_perturbation_MOD_s_superposition_instability_wave
	at /p/work/hyeoksu/MFC/lsa/cpu-debug/build/staging/5cfdab4a72/fypp/pre_process/m_perturbation.fpp.f90:131
#7  0x4778d2 in __m_initial_condition_MOD_s_generate_initial_condition
	at /p/work/hyeoksu/MFC/lsa/cpu-debug/build/staging/5cfdab4a72/fypp/pre_process/m_initial_condition.fpp.f90:310
#8  0x4d36a7 in __m_start_up_MOD_s_apply_initial_condition
	at /p/work/hyeoksu/MFC/lsa/cpu-debug/build/staging/5cfdab4a72/fypp/pre_process/m_start_up.fpp.f90:843
#9  0x4374e4 in p_main
	at /p/work/hyeoksu/MFC/lsa/cpu-debug/src/pre_process/p_main.f90:36
#10  0x4375b6 in main
	at /p/work/hyeoksu/MFC/lsa/cpu-debug/src/pre_process/p_main.f90:11

mfc: ERROR > :( /p/work/hyeoksu/MFC/lsa/cpu-debug/build/install/5cfdab4a72/bin/pre_process failed with exit code 2.
```

I don't know why it didn't appear in other compilations. Also it is interesting that I got the expected results from them..! 

Anyway it should be okay now.
